Add of Online Hierarchical Clustering

kchardon opened the pull request:
<!--
READ ME!

Thanks for contributing to River!

If you're new to the project, then we encourage you to first [open a discussion](https://github.com/online-ml/river/discussions/new). This helps everyone save time by making sure we're all aligned on the contribution that is being made.

Have a great day.
-->

kchardon committed:
add of Hierarchical Clustering

kchardon committed:
Fixed issues

kchardon committed:
Fix issues

kchardon committed:
Fixed issues

kchardon committed:
Fix issues

kchardon committed:
fixed trailing spaces

kchardon committed:
Fixed black and isort

kchardon committed:
reverting to the version passing the 'build river ubuntu'

kchardon committed:
fixed isort and black

MaxHalford commented:
Hey there! I hope it's ok for me to answer only by now.

Am I correct in assuming the algorithm stores all the data points it sees in memory (i.e. the `X` attribute)?

kchardon commented:
> Hey there! I hope it's ok for me to answer only by now.
> 
> Am I correct in assuming the algorithm stores all the data points it sees in memory (i.e. the `X` attribute)?

Hello, sorry for replying that late.
I use the `window_size` attribute and when there are more data than allowed, it deletes the oldest data point.
If `window_size` equals 0, then it stores all the data points

kchardon commented:
> > If window_size < 1, then it stores all the data points
> 
> Ok I see, fair. But I don't think we'll ever want that behavior. Could you remove it?

Yes I can. So I add an error if the value of `window_size` is not an integer > 0 ?

MaxHalford commented:
Nope, no need to check for an error. An exception will raise itself at some point. In general, we don't do input validation. Instead, we document well.

kchardon commented:
Okay ! I will delete it

kchardon commented:
If I delete the line, if the user put a window_size < 1, it will automatically store all the data points, because I do 

if len(self.X.keys()) >= self.window_size: delete oldest node and replace it by the new node
else : add the node as a new one

So since len(self.X.keys()) is always > 0, it will add all the nodes and use all the data points.

Do I replace the else with elif len(self.X.keys()) < self.window_size : add the node as a new one

With that it will raise an error if the window size is not > 0

kchardon committed:
Deleted the possibility to use all the data points

kchardon committed:
correction for ruff hook

kchardon committed:
correction for ruff hook

kchardon committed:
correction for ruff hook

hoanganhngo610 committed:
Modify Eucliean distance calculation using np.linalg.norm for better computational speed.

hoanganhngo610 committed:
Refactor elements related to the distance function.

hoanganhngo610 committed:
Remove data types of attributes

hoanganhngo610 committed:
Refactor inter subtree similarity function.

hoanganhngo610 committed:
Refactor intra subtree similarity function

hoanganhngo610 committed:
Refactor leave finding function.

hoanganhngo610 committed:
Refactor online top down (OTD) clustering function.

hoanganhngo610 committed:
Refactor.

hoanganhngo610 committed:
Cheng function name OTD to otd_clustering.

hoanganhngo610 commented:
Hi @kchardon! I am Hoang-Anh, the maintainer of the clustering module of River. For now, I think that it would be best if I can take over the review of this PR and proceed with it to get it merged to River as soon as possible, since from my first glance, the PR has been of really high quality.
If you wouldn't mind, first, I will start by refactoring the code and make changes to make it align with the rest from River. After all necessary changes are in place, I will get @MaxHalford to have a final look and give the green light.

hoanganhngo610 committed:
Remove unnecessary comments.

hoanganhngo610 committed:
Refactor description of the algorithm within Docstring.

hoanganhngo610 committed:
Refactor tests in Docstring.

hoanganhngo610 committed:
Refactor merge_nodes function.

hoanganhngo610 committed:
Refactor comments in merge_nodes

hoanganhngo610 committed:
Rename predict_otd.

hoanganhngo610 committed:
Simplify comments.

hoanganhngo610 committed:
Simplify comments.

hoanganhngo610 committed:
Modify __str__ printed output by adding "Printed Hierarchical Clustering Tree." at the end of the tree.

hoanganhngo610 committed:
Modify __str__ printed output by adding "Printed Hierarchical Clustering Tree." at the end of the tree.

hoanganhngo610 committed:
Rename predict_otd.

hoanganhngo610 committed:
Rename predict_otd.

hoanganhngo610 committed:
Split comments and rename printTree to print_tree.

hoanganhngo610 committed:
Split comments and rename printTree to print_tree.

hoanganhngo610 committed:
Modify self.X to self.x_clusters.

hoanganhngo610 committed:
Modify self.X to self.x_clusters.

hoanganhngo610 committed:
Lexical changes.

hoanganhngo610 committed:
Lexical changes.

hoanganhngo610 committed:
Remove unnecessary comments.

hoanganhngo610 committed:
Remove unnecessary comments.

MaxHalford committed:
Support pandas v2 (#1321)

hoanganhngo610 committed:
Refactor Docstring

hoanganhngo610 committed:
Refactor Docstring

hoanganhngo610 committed:
Refactor comment.

hoanganhngo610 committed:
Refactor comment.

hoanganhngo610 committed:
Make find_path() a static method.

hoanganhngo610 committed:
Make find_path() a static method.

hoanganhngo610 committed:
Refactor Docstring.

hoanganhngo610 committed:
Refactor Docstring.

hoanganhngo610 committed:
Make print_tree static method.

hoanganhngo610 committed:
Make print_tree static method.

hoanganhngo610 committed:
Refactor code to account for failing tests.

hoanganhngo610 committed:
Refactor code to account for failing tests.

hoanganhngo610 committed:
Refactor distance function used in Hierarchical Clustering class.

hoanganhngo610 committed:
Refactor distance function used in Hierarchical Clustering class.

hoanganhngo610 committed:
Delete euclidean_distance function (due to being unnecessary).

hoanganhngo610 committed:
Delete euclidean_distance function (due to being unnecessary).

hoanganhngo610 committed:
Code refactoring to align with other algorithms available in River.

hoanganhngo610 committed:
Code refactoring to align with other algorithms available in River.

hoanganhngo610 committed:
Modify Docstring description for dist_func.

hoanganhngo610 committed:
Modify Docstring description for dist_func.

hoanganhngo610 commented:
@MaxHalford I think I'm quite happy with the current state of the code! Only one small concern is that the unit tests keep failing because there is an attribute error, saying that `HierarchicalClustering` does not have the attribute `distance_func`, although I believe atm there is no `distance_func` in my code anywhere anymore. Do you think of any potential reasons behind this?

hoanganhngo610 committed:
Delete least common ancestor finding function (since this function is not used within the algorithm).

hoanganhngo610 committed:
Delete least common ancestor finding function (since this function is not used within the algorithm).

smastelini committed:
Simplify Adaptive Random Forest (#1323)

hoanganhngo610 commented:
@MaxHalford @kchardon If possible, I would really hope that both of you would be able to take a final look at the current state of the implementation and see if there are any changes you want to make. In case there is no opposition from both of you, when applicable, I would want to merge this into the main branch of `River`.

MaxHalford commented:
I will review next week :)

MaxHalford committed:
remove pytest from gaussian.py

MaxHalford committed:
Update test_gaussian.py

jacowp357 committed:
added web traffic dataset and info (#1326)

* added web traffic dataset and info

* removed CSV file and updated the url

MaxHalford committed:
tidy up WebTraffic

MaxHalford committed:
solve identical samples issue in lof

MaxHalford commented:
I (finally) took a look at this. There's a lot of nits to fix. But the main issue I see is that the code relies on numpy. It converts input dictionaries to numpy arrays. Would it be possible to use dictionaries only instead? I don't see any good reason to rely on numpy. Indeed, it's not in the spirit of River to rely on numpy when it can be avoided.

MaxHalford committed:
Update delete-caches.yml

MarekWadinger committed:
Add _from_state methods to MultivariateGaussian and EmpiricalCovariance (#1327)

* ADD: _from_state method for EmpiricalCovariance

* ADD: _from_state method to MultivariateGaussian

* ADD: docstring to _from_state method

* UPDATE: add _from_state info to release notes

* ADD: doctest on _from_state method

* FIX: end of file

* FIX: resolve comments from smastelini

smastelini committed:
expand ADWIN docs (#1334)

smastelini committed:
Expand tree.ExtremelyFastDecisionTree documentation

Commit:
pre-commit actions

MaxHalford committed:
update installation notes

smastelini committed:
Benchmarks revamp (#1335)

* tidy-up benchmarks

* update benchmark dependencies

Commit:
update benchmark files

hoanganhngo610 committed:
DBSTREAM fix (related to issue #1324) (#1336)

* Adding negative signs before fading_factor for steps within Algoritjm 2 of the paper by Hashler and Bolanos to allow clusters with low weight removed.

* Change clustering_is_up_to_date to True after every time the function recluster is called.

* initiate new micro cluster based on the maximum key of the existing micro clusters, or indexed as 0 if the list of micro clusters is still empty.

* Add description to the UNRELEASED.md file

MarekWadinger committed:
FIX: LOF with QuantileFilter raises IndexError (#1330)

* FIX: LOF with QuantileFilter raises IndexError

* FIX: score_one changes internal state of LOF

* FIX: black refactor

* Update initial score in LOF

Co-authored-by: Max Halford <maxhalford25@gmail.com>

* Add tests for stateless score_one and fix doctest

* Minor line length fix

* Add LOF fixtures to release notes

---------

Co-authored-by: Max Halford <maxhalford25@gmail.com>

smastelini committed:
Fixes #1248 (#1337)

* bug fix

* update affected test

donny741 committed:
DBSTREAM: fix density graph timestamps (#1338)

donny741 committed:
DBSTREAM: fix microcluster labeling (#1339)

* Fix microcluster labeling

* Add tests

* Update unreleased.md

* Fix linter errors

* Fix imports order

donny741 committed:
dbstream: fix adjacency matrix building (#1340)

* dbstream: fix adj matrix building

* Update unreleased.md

* Update docs/releases/unreleased.md

Co-authored-by: Max Halford <maxhalford25@gmail.com>

---------

Co-authored-by: Max Halford <maxhalford25@gmail.com>

MaxHalford committed:
Split events into timestamps and datum (#1345)

MaxHalford committed:
Remove dict2numpy and numpy2dict (#1344)

* remove dict2numpy and numpy2dict

* Update unreleased.md

MaxHalford committed:
fix test

smastelini committed:
Fix bug in EFDT (#1347)

smastelini committed:
Add the NoDrift dummy drift detector (#1350)

* add NoDrift detector

* lint

* update test

* Update river/drift/no_drift.py

Co-authored-by: Max Halford <maxhalford25@gmail.com>

* Update river/drift/no_drift.py

Co-authored-by: Max Halford <maxhalford25@gmail.com>

* Update river/drift/no_drift.py

Co-authored-by: Max Halford <maxhalford25@gmail.com>

* streamline test

---------

Co-authored-by: Max Halford <maxhalford25@gmail.com>

MaxHalford committed:
add yield_predictions to iter_progressive_val_score

MaxHalford committed:
Merge branch 'main' of https://github.com/online-ml/river

MaxHalford committed:
Update progressive_validation.py

hoanganhngo610 committed:
Implementation of the incremental Kolmogorov-Smirnov Statistics (#1354)

* Initial implementation of incremental KS statistics

* Update docstring

* Refactor incremental KS from metrics to stats.

* Modify import in test_ks.py file

* Refactor import in test_ks.py file

* Refactor import for test_ks.py

* Add test to assert whether the two sliding windows follow the same distribution with the result from the KS test.

* Run black code formatter on kolmogorov_smirnov.py file

* Fix test_ks.py with ruff v0.1.2

* Commit from suggestion of code reviewer.

Co-authored-by: Max Halford <maxhalford25@gmail.com>

* Update code from reviewer's suggestion.

Co-authored-by: Max Halford <maxhalford25@gmail.com>

* Update code from reviewer's suggestion.

Co-authored-by: Max Halford <maxhalford25@gmail.com>

* Trim trailing whitespace of kolmogorov_smirnov.py file

* Add description to UNRELEASED.md

---------

Co-authored-by: Max Halford <maxhalford25@gmail.com>

MaxHalford committed:
tidy ks

hoanganhngo610 committed:
Implementation of the StandardAbsoluteDeviation (SAD) anomaly detection algorithm (#1357)

Implementation of the SAD algorithm, adapted from the version implemented in PySAD (Python Streaming Anomaly Detection).

MaxHalford committed:
Use Poetry (#1346)

* start using poetry

* Update pyproject.toml

* update poetry

* Update __version__.py

* Update unit-tests.yml

* make groups optional

* no need to source activate

* Update __version__.py

* Update __version__.py

* bumps

* wip

* wip

* add pytest-xdist[psutil]

* Update pyproject.toml

* Update build.py

* wip

* Update action.yml

* Update unit-tests.yml

* Update unit-tests.yml

* Update unit-tests.yml

* Update unit-tests.yml

* Update unit-tests.yml

* Update build.py

* try this

* now try this

* Update unit-tests.yml

* Update sad.py

* Update sad.py

* fix tests

* stuff

* Update action.yml

* Update CONTRIBUTING.md

* Update action.yml

* try adding a build-root param

* wip

* Update unit-tests.yml

* wip

* Update action.yml

* wip

* Update unit-tests.yml

MaxHalford committed:
merge

dependabot[bot] committed:
Bump werkzeug from 3.0.0 to 3.0.1 (#1438)

Bumps [werkzeug](https://github.com/pallets/werkzeug) from 3.0.0 to 3.0.1.
- [Release notes](https://github.com/pallets/werkzeug/releases)
- [Changelog](https://github.com/pallets/werkzeug/blob/main/CHANGES.rst)
- [Commits](https://github.com/pallets/werkzeug/compare/3.0.0...3.0.1)

---
updated-dependencies:
- dependency-name: werkzeug
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

MaxHalford committed:
Update README.md

MaxHalford committed:
Merge branch 'main' of https://github.com/online-ml/river

MaxHalford committed:
ellipsis

MaxHalford committed:
mmm

MaxHalford committed:
Update README.md

MaxHalford committed:
Update README.md

MaxHalford committed:
add ipykernel

MaxHalford committed:
Merge branch 'main' of https://github.com/online-ml/river

MaxHalford committed:
Make confusion matrix work with evaluate module (#1444)

* make cm work with evaluate

* Update test_confusion.py

MaxHalford committed:
bump to 0.20

MaxHalford committed:
fix

MaxHalford committed:
Update release-docs.yml

MaxHalford committed:
Update pypi.yml

MaxHalford committed:
Update release-docs.yml

MaxHalford committed:
wip

MaxHalford committed:
Update release-docs.yml

MaxHalford committed:
update poetry

MaxHalford committed:
Update pypi.yml

MaxHalford committed:
Update Makefile

MaxHalford committed:
add jupyter

MaxHalford committed:
Update release-docs.yml

MaxHalford committed:
upgrade cibuildwheel

MaxHalford committed:
Update release-docs.yml

MaxHalford committed:
wip

MaxHalford committed:
Update pypi.yml

MaxHalford committed:
Update release-docs.yml

MaxHalford committed:
0.20.1

MaxHalford committed:
Update README.md

MaxHalford committed:
make cm inherit from multi-class metric

mariliatd committed:
Rename sample_weight to w (code quality) (#1457)

* rename sample_weight to w in cluster module files

Signed-off-by: mariliatd <mariliatd@gmail.com>

* rename sample_weight to w in ensemble module files

Signed-off-by: mariliatd <mariliatd@gmail.com>

* rename sample_weight to w in facto module files

Signed-off-by: mariliatd <mariliatd@gmail.com>

* rename sample_weight to w in forest and tree modules files

Signed-off-by: mariliatd <mariliatd@gmail.com>

* rename sample_weight to w in metrics module files

Signed-off-by: mariliatd <mariliatd@gmail.com>

* rename sample_weight to w in tree splitter module files

Signed-off-by: mariliatd <mariliatd@gmail.com>

* add note to unreleased.md file

Signed-off-by: mariliatd <mariliatd@gmail.com>

---------

Signed-off-by: mariliatd <mariliatd@gmail.com>

MaxHalford committed:
wip

MaxHalford committed:
Merge branch 'main' of https://github.com/online-ml/river

MaxHalford committed:
ruff

MaxHalford committed:
move stuff around

MaxHalford committed:
Update pypi.yml

MaxHalford committed:
Update pypi.yml

MaxHalford committed:
Update pypi.yml

MaxHalford committed:
Update pypi.yml

MaxHalford committed:
Update pypi.yml

MaxHalford committed:
Update pypi.yml

MaxHalford committed:
Update pypi.yml

MaxHalford committed:
Update pypi.yml

MaxHalford committed:
Update pypi.yml

MaxHalford committed:
Update pypi.yml

MaxHalford committed:
Update pypi.yml

MaxHalford committed:
Update pypi.yml

MaxHalford committed:
wip

MaxHalford committed:
Update pypi.yml

mariliatd committed:
No more return self (#1461)

* remove return self from learn_one of active learning

Signed-off-by: mariliatd <mariliatd@gmail.com>

* remove return self from learn_one of cluster module

Signed-off-by: mariliatd <mariliatd@gmail.com>

* remove return self from learn_one of factorization machines classes

Signed-off-by: mariliatd <mariliatd@gmail.com>

* remove return self from learn_one of forest classes

Signed-off-by: mariliatd <mariliatd@gmail.com>

* remove return self from learn_one of ranking models classes

Signed-off-by: mariliatd <mariliatd@gmail.com>

* remove return self from learn_one of tree models classes

Signed-off-by: mariliatd <mariliatd@gmail.com>

* remove return self from learn_one of mondrian tree models classes

Signed-off-by: mariliatd <mariliatd@gmail.com>

* remove return self from learn_one of compose module

Signed-off-by: mariliatd <mariliatd@gmail.com>

* remove return self from learn_one of preprocessing module

Signed-off-by: mariliatd <mariliatd@gmail.com>

* remove return self from learn_one of anomaly detector classes

Signed-off-by: mariliatd <mariliatd@gmail.com>

* remove return self from learn_one of base module

Signed-off-by: mariliatd <mariliatd@gmail.com>

* remove return self from learn_one of feature extraction module

Signed-off-by: mariliatd <mariliatd@gmail.com>

* remove return self from learn_one of linear models classes

Signed-off-by: mariliatd <mariliatd@gmail.com>

* remove return self from learn_one of naive bayes models classes

Signed-off-by: mariliatd <mariliatd@gmail.com>

* remove return self from learn_one of remaining files

Signed-off-by: mariliatd <mariliatd@gmail.com>

* remove return self from learn_one of remaining files

Signed-off-by: mariliatd <mariliatd@gmail.com>

* remove return self from learn_many of all classes

Signed-off-by: mariliatd <mariliatd@gmail.com>

* remove unnecessary learn_one assignment to variable in jupyter notebooks

Signed-off-by: mariliatd <mariliatd@gmail.com>

---------

Signed-off-by: mariliatd <mariliatd@gmail.com>

MaxHalford committed:
Update pypi.yml

MaxHalford committed:
Merge branch 'main' of https://github.com/online-ml/river

MaxHalford committed:
remove return self from update and revert

MaxHalford committed:
Update emp.py

MaxHalford committed:
bump

MaxHalford committed:
Update base.py

MaxHalford committed:
Update no_drift.py

MaxHalford committed:
Update no_drift.py

MaxHalford committed:
fix examples

MaxHalford committed:
Update __version__.py

MaxHalford committed:
fix adwin

hoanganhngo610 committed:
Remove math.sqrt from the distance calculating function within DBSTREAM (resolve partially issue #1468).

dependabot[bot] committed:
Bump jupyter-server from 2.11.1 to 2.11.2 (#1469)

hoanganhngo610 committed:
Modify the distance calculating function in Silhouette coefficient (related to an issue raised in #1468).

hoanganhngo610 committed:
Modify the distance calculating function in DenStream (relating to an issued mentioned in #1468).

hoanganhngo610 committed:
Change the inserted value of the shared density when KeyError occurs to 1 instead of 0.

hoanganhngo610 committed:
Removing entries within shared density and associated timestamp when removing a micro-cluster (related to an inquiry from #1468)

hoanganhngo610 committed:
Modify the first test within test_dbstream.py file

hoanganhngo610 committed:
Update the second test within test_dbstream.py file

hoanganhngo610 committed:
Add test for DBSTREAM with synthetic data

hoanganhngo610 committed:
Update test_density_graph_with_removed_microcluster function in DBSTREAM testing

hoanganhngo610 committed:
Modify test_density_graph_with_three_micro_clusters in DBSTREAM test file

hoanganhngo610 committed:
Update test_dbstream.py file after pre commit checks.

mbispham committed:
Add license to riverml.xyz (#1480)

MaxHalford committed:
Update airline_passengers.py

MaxHalford committed:
Merge branch 'main' of https://github.com/online-ml/river

gabrieljaguiar committed:
Add new drift detectors (#1474)

* fhddm

* fhddm

* fhddm_s

* annotations

* code quality fixes

* code quality fix

* code quality

* codequality

* doc fixes and private variables

* doc fixes

* using collections.deque

* merge detectors

* docstring and merge detectors

* fixes and merge

* fhddms added to example

* test fixes

* fixes

* fixes

* fixing example

* example fix

* unreleased file

* remove count

* remove count

niccolopetti committed:
Add stream.iter_polars #1503 (#1504)

MarekWadinger committed:
Fix missing attribute  '_from_state' in 'EmpiricalCovariance' (#1471)

* ADD: from_state allows int as covariance

* FIX: emp _from_state

* FIX: broken statefulness of update_many

* FIX: remove unexpected kwarg

* FIX: if statement nesting + redundant docstring entry

* FIX: wrong intendation + sort keys for shuffled samples

AdilZouitine committed:
Fix build from source (sdist) (#1506)

* Fix build from source (sdist)

* Delete MANIFEST.in

smastelini committed:
Fix issue #1507 (SWINN streamline) (#1508)

* streamline code

* now I remember why those methods exist

* change test

* release notes

niccolopetti committed:
add a trained_on_dist to imblearn/random.py #1511 (#1513)

sebiwtt committed:
Adding PredictiveAnomalyDetection to river/anomaly (#1458)

Add Predictive Anomaly Detection, a semi-supervised technique to anomaly detection that employs a predictive model to learn the behavior of a dataset.

hoanganhngo610 committed:
Add PAD to unreleased.md file.

MaxHalford committed:
Update __version__.py

MaxHalford committed:
Merge branch 'main' of https://github.com/online-ml/river

MaxHalford committed:
prepare release

MaxHalford committed:
Update pyproject.toml

FBruzzesi committed:
pin version (#1525)

FBruzzesi committed:
docs: update Pipelines recipe (#1528)

- Document was out of date wrt. the introduction of the
`compose.learn_during_predict` context manager.

MaxHalford committed:
try auto

MaxHalford committed:
Merge branch 'main' of https://github.com/online-ml/river

MaxHalford committed:
Update pypi.yml

MaxHalford committed:
Update pypi.yml

MaxHalford committed:
Update pypi.yml

dependabot[bot] committed:
Bump pillow from 10.1.0 to 10.3.0 (#1520)

Bumps [pillow](https://github.com/python-pillow/Pillow) from 10.1.0 to 10.3.0.
- [Release notes](https://github.com/python-pillow/Pillow/releases)
- [Changelog](https://github.com/python-pillow/Pillow/blob/main/CHANGES.rst)
- [Commits](https://github.com/python-pillow/Pillow/compare/10.1.0...10.3.0)

---
updated-dependencies:
- dependency-name: pillow
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

PabloRuizCuevas committed:
Update pyproject.toml (#1533)

* Update pyproject.toml

Changed python requirements, if the future constrain is there other project with less strict requirements will fail to install river,

for instance i have in my pyproject.toml python = "^3.11" and fails with "requires Python <3.13,>=3.9, river is forbidden."

* dependencies update

Commit:
fix failing test (wrt #1253)

smastelini committed:
Adding ODAC (#1537)

* add ODAC algorithm

* ODAC passed in pytest

* draw() fixed

* multiple fixes

* fix docs

* mypy

* Changes per Gonçalo's review

* fix failing test (wrt #1253)

* fix tests

* nitpick

* more nitpick

* release notes

---------

Co-authored-by: Bezum30 <goncalo.augusto@icloud.com>

stereosky committed:
Update docs for typos and wording (#1544)

Commit:
ODAC: quality of life improvements (sent by Gonçalo)

gbolmier committed:
Rename release 0.21.1 release notes to 0.21.1.md (#1546)

gbolmier committed:
Make polars an optional dependency (#1545)

- Fix #1504 that, by mistake, added polars as a required
dependency when adding stream.iter_polars.
- Run `poetry lock`.

smastelini committed:
Fix #1560 (ARF cornercase)

e10e3 committed:
Upgrade the pyo3 crate (#1564)

Fixes https://github.com/online-ml/river/discussions/1563

e10e3 committed:
Fix mypy errors (#1562)

Mark the parameters with a default value of None as optionals.

gbolmier committed:
Release 0.21.2

gbolmier committed:
Downgrade release-docs action Py version to 3.11

Temporary downgrade the Python version to 3.11 in
the release-docs workflow to avoid a Pydantic
error when importing spacy.

TypeError: ForwardRef._evaluate() missing 1
required keyword-only argument: 'recursive_guard'

gbolmier committed:
Revert "Downgrade release-docs action Py version to 3.11"

This reverts commit 20ef984585d8794167572b223b946cbb91271b07.

gbolmier committed:
Update install-env and release-docs workflows (#1569)

- Update actions to latest versions
- Update Python version to 3.12.3 in release-docs workflow
- Increment Poetry installation cached key in install-env workflow

gbolmier committed:
Update PyPI workflow (#1571)

* Update pypa/cibuildwheel in pypi.yml

* Remove ppc64le and s390x builds from pypi.yml

RavSS committed:
Fixed `AttributeError` in metrics/roc_auc.py. (#1568)

`scipy.integrate.trapezoid` replaced `scipy.integrate.trapz` in SciPy.

e10e3 committed:
Replace Gym with Gymnasium (#1580)

* Replace Gym with Gymnasium

The maintainers of OpenAI's Gym decided to stop maintaining the library
and moved over to a fork called Gymnasium.

They advise to switch to the new library, which can be used as a drop-in
replacement.

See: https://github.com/openai/gym/blob/master/README.md

* Upgrade Gymnasium

* Add a float conversion to please MyPy

e10e3 committed:
Clear out MyPy's warnings (#1581)

Type annotations are added to some functions; this allows to remove
MyPy's warnings when typechecking the code by clearing some
inconsistencies.

A couple incorrect or extraneous annotations were removed in the
process.

dependabot[bot] committed:
Bump urllib3 from 2.2.1 to 2.2.2 (#1557)

Bumps [urllib3](https://github.com/urllib3/urllib3) from 2.2.1 to 2.2.2.
- [Release notes](https://github.com/urllib3/urllib3/releases)
- [Changelog](https://github.com/urllib3/urllib3/blob/main/CHANGES.rst)
- [Commits](https://github.com/urllib3/urllib3/compare/2.2.1...2.2.2)

---
updated-dependencies:
- dependency-name: urllib3
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump setuptools from 69.5.1 to 70.0.0 (#1575)

Bumps [setuptools](https://github.com/pypa/setuptools) from 69.5.1 to 70.0.0.
- [Release notes](https://github.com/pypa/setuptools/releases)
- [Changelog](https://github.com/pypa/setuptools/blob/main/NEWS.rst)
- [Commits](https://github.com/pypa/setuptools/compare/v69.5.1...v70.0.0)

---
updated-dependencies:
- dependency-name: setuptools
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

e10e3 committed:
Use units based on powers of 2 (#1583)

* Use units based on powers of 2

This reflects how they are interpreted in the code.

* Add a changelog entry

Commit:
[docs] fix memory units in the tree guideline

dependabot[bot] committed:
Bump zipp from 3.18.2 to 3.19.1 (#1572)

Bumps [zipp](https://github.com/jaraco/zipp) from 3.18.2 to 3.19.1.
- [Release notes](https://github.com/jaraco/zipp/releases)
- [Changelog](https://github.com/jaraco/zipp/blob/main/NEWS.rst)
- [Commits](https://github.com/jaraco/zipp/compare/v3.18.2...v3.19.1)

---
updated-dependencies:
- dependency-name: zipp
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump certifi from 2024.2.2 to 2024.7.4 (#1567)

Bumps [certifi](https://github.com/certifi/python-certifi) from 2024.2.2 to 2024.7.4.
- [Commits](https://github.com/certifi/python-certifi/compare/2024.02.02...2024.07.04)

---
updated-dependencies:
- dependency-name: certifi
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

e10e3 committed:
Limit tree growth to the system recursion limit (#1585)

Commit:
[docs] remove unavailable tree algorithm and fix tree depth behavior description in the tree guidelines

bezumbzalinho committed:
Updates in ODAC code (#1584)

* update description, fix a bug when calculating rnomc, add draw() method to see hierarchical cluster's structure as a Graphviz graph, and working with Var when cluster has only one time-series

* change the description of draw() method

* add cluster's name in draw() method

* correct version

* Update river/cluster/odac.py

* Update river/cluster/odac.py

* update docs/releases/unreleased.md

* Update docs/releases/unreleased.md

---------

Co-authored-by: gonfa3003 <116749870+gonfa3003@users.noreply.github.com>
Co-authored-by: Saulo Martiello Mastelini <saulomastelini@gmail.com>

e10e3 committed:
Remove mention of former compat modules (#1588)

* Remove mention of former `compat` modules

* Format test_estimators.py

e10e3 committed:
Remove script download from polyfill.io (#1590)

The domain polyfill.io was taken over by a malicious actor to download
malware.
Polyfill is not needed for modern browsers.

It is recommended to remove references to the website.

See: https://blog.qualys.com/vulnerabilities-threat-research/2024/06/28/polyfill-io-supply-chain-attack

e10e3 committed:
Remove the class CentralMoments (#1593)

* Remove the class CentralMoments

This was a utility class internal to River, and the last references to
it were removed in a previous version of the library, with the move to
Rust statistics.

* Add a changelog entry

e10e3 committed:
Add type stubs for Cython files (#1594)

* Check for missing imports in River

* Add type stubs for expected_mutual_info.pyx

* Add type stubs for efficient_rollingrocauc.pyx

* Add type stub for adwin_c

* Add basic type stubs for vectordict

* Remove modulo from Vectordict.__pow__

This brings the signature in line with how Python defines the operator.
The modulo parameter was not used.

* Check the stubs files in pre-commit

MaxHalford committed:
Bump dev tools (#1599)

* Bump dev tools

* Update poetry.lock

Daveismus committed:
Add Hack for stats.IQR get method for avoiding panicking (#1603)

Co-authored-by: David Brielbeck <david.brielbeck@schufa.de>

dependabot[bot] committed:
Bump notebook from 7.2.0 to 7.2.2 (#1604)

Bumps [notebook](https://github.com/jupyter/notebook) from 7.2.0 to 7.2.2.
- [Release notes](https://github.com/jupyter/notebook/releases)
- [Changelog](https://github.com/jupyter/notebook/blob/@jupyter-notebook/tree@7.2.2/CHANGELOG.md)
- [Commits](https://github.com/jupyter/notebook/compare/@jupyter-notebook/tree@7.2.0...@jupyter-notebook/tree@7.2.2)

---
updated-dependencies:
- dependency-name: notebook
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump jupyterlab from 4.2.0 to 4.2.5 (#1605)

Bumps [jupyterlab](https://github.com/jupyterlab/jupyterlab) from 4.2.0 to 4.2.5.
- [Release notes](https://github.com/jupyterlab/jupyterlab/releases)
- [Changelog](https://github.com/jupyterlab/jupyterlab/blob/@jupyterlab/lsp@4.2.5/CHANGELOG.md)
- [Commits](https://github.com/jupyterlab/jupyterlab/compare/@jupyterlab/lsp@4.2.0...@jupyterlab/lsp@4.2.5)

---
updated-dependencies:
- dependency-name: jupyterlab
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

e10e3 committed:
Type hints for 'return self' (#1601)

* Make all learn and update methods return None

Changing the return types in the base classes means MyPy can detect when
a method returns an unexpected value.

* Propagate None returns to child classes

* Remove wrong returns

* Add a note in the changelog

* Change the return type of SortedWindow.append

* Fix new MyPy 11 errors

mbispham committed:
Fixup typo (#1606)

dependabot[bot] committed:
Bump actions/download-artifact from 3 to 4.1.7 in /.github/workflows (#1607)

Bumps [actions/download-artifact](https://github.com/actions/download-artifact) from 3 to 4.1.7.
- [Release notes](https://github.com/actions/download-artifact/releases)
- [Commits](https://github.com/actions/download-artifact/compare/v3...v4.1.7)

---
updated-dependencies:
- dependency-name: actions/download-artifact
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

danielnowakassis committed:
Adding Local Adaptive Streaming Tree (#1610)

* add first last with out-of-date docs

* update LAST to detect change in the data distribution + iter_arff none class

Docs are also updated

* update docs

* Update hoeffding_adaptive_tree_classifier.py

* changes after tests

* solving inheritance and small fixes

* tests + current_merit method

* Update river/tree/hoeffding_adaptive_tree_classifier.py

* update docs

* change docs

* change last

* Update docs/releases/unreleased.md

* Update river/tree/hoeffding_adaptive_tree_classifier.py

* Update river/tree/last_classifier.py

* add disclamer

* Update river/tree/last_classifier.py

---------

Co-authored-by: Saulo Martiello Mastelini <saulomastelini@gmail.com>

denisesato committed:
Bugfix for detecting change points based on MOA (#1613)

* Bugfix based on MOA source code in the _detect_change method. The k index should increment from 0 to bucket.current_idx - 1. The previous code "for k in range(bucket.current_idx - 1):" only increment k to bucket.current_idx - 2 because of the range function.

* Bugfix based on MOA source code in the _detect_change method. The k index should increment from 0 to bucket.current_idx - 1. The previous code "for k in range(bucket.current_idx - 1):" only increment k to bucket.current_idx - 2 because of the range function.

* Tests updated due to bugfix on the ADWIN change detector - reported on issue #1614

smastelini committed:
Update release notes

smastelini committed:
Update release notes

e10e3 committed:
New MyPy configuration for per-module type completeness (#1596)

* Make River fully type-checked by default

Add an override for all modules while they are not annotated.

* Use the true location of Gymnasium's registry

The registry variable of Gymnasium environments is located in the
'registration' module.
This variables is also imported by the 'envs' module, but not explictely
re-exported.
This means the previous code was correct at runtime, but MyPy raised
easily-corrected warnings.

* Enable more rules for all modules

These rules do not cause errors in the current code, meaning they can be
enabled without harm.

MaxHalford committed:
# type: ignore[misc]

gbolmier committed:
Update Py version to 3.12.X in release-docs.yml

Previous cap to 3.12.3 should not be needed anymore since
explosion/spaCy#13550 seems to be fixed

kchardon committed:
removing use of numpy

kchardon committed:
removing use of numpy

kchardon committed:
Merge branch 'main' of https://github.com/kchardon/river_project

kchardon committed:
ruff format

kchardon committed:
odac issue

kchardon committed:
remove hcluster

kchardon committed:
last main commit

kchardon committed:
hcluster

kchardon committed:
ruff

kchardon committed:
new test

kchardon committed:
correct test

kchardon commented:
Hello, sorry to come back to this PR so late.
I've changed the code to not rely on numpy and fixed the issues during the checks.
Let me know if anything else can be improved!

MaxHalford commented:
Closing because the PR is messed up. But happy to pick back up the reviews if there's still some will!
